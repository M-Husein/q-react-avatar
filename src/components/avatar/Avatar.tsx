import '../../style/components/Avatar.scss';
import type { CSSProperties } from 'react';
import { useState, forwardRef } from 'react';
import { joinClasses, str2Hex, getInitials, darkOrLight } from '../../utils';
import { parseSize } from './utils';

export interface AvatarProps {
  src?: string | any
  alt?: string
  title?: string
  size?: number | "xs" | "sm" | "md" | "lg"
  style?: CSSProperties
  loading?: undefined | 'lazy' | 'eager' // eager | lazy (Default browser = eager)
  decoding?: undefined | 'async' | 'sync' | 'auto' // sync | async | auto (Default browser = auto)
  draggable?: boolean
  className?: string
  bg?: string
  onError?: (e: any) => void
  onLoad?: (e: any) => void
}

export const Avatar = forwardRef<HTMLImageElement, any>(
  (
    {
      size = 'md',
      src,
      alt = ' ',
      title,
      style, 
      loading = 'lazy',
      decoding = 'async',
      draggable = false,
      className,
      bg, // DEFINE background-color
      onError,
      onLoad,
      ...etc
    }: AvatarProps,
    ref
  ) => {
    let errStyle = {},
        initName,
        errClass = '',
        trm = alt.trim(),
        color = '',
        initLoad = true;

    const fixSize = parseSize(size);

    const setDarkLight = (c: string) => darkOrLight(c) === 'dark' ? 'ava-light' : 'ava-dark';

    if(!src){
      color = bg ? bg.replace('#', '') : str2Hex(trm);
      errStyle = {
        '--fs': 'calc(' + fixSize + 'px / 2.25)',
        '--bg': '#' + color
      };
      initName = getInitials(trm);
      errClass = setDarkLight(color);
      initLoad = false;
    }

    const [load, setLoad] = useState<boolean>(initLoad);
    const [errorStyle, setErrorStyle] = useState(errStyle);
    const [initial, setInitial] = useState(initName);
    const [errorClass, setErrorClass] = useState<string>(errClass);

    const Load = (e: React.SyntheticEvent<HTMLImageElement, Event>) => {
      setLoad(false);
      setInitial(undefined);
      setErrorClass('');
      setErrorStyle({});
      onLoad?.(e);
    };

    const Err = (e: React.SyntheticEvent<HTMLImageElement, Event>) => {
      setLoad(false);

      if(!initial){
        color = bg ? bg.replace('#', '') : str2Hex(trm);
  
        setInitial(getInitials(trm));
        setErrorClass(setDarkLight(color));
        setErrorStyle({
          '--fs': 'calc(' + fixSize + 'px / 2.25)',
          '--bg': '#' + color
        });
      }

      onError?.(e);

      return;
    };

    return (
      <img
        {...etc}
        ref={ref}
        decoding={decoding}
        loading={loading}
        width={fixSize}
        height={fixSize} 
        alt={alt}
        src={src}
        aria-label={initial}
        title={title}
        style={{
          ...errorStyle, 
          ...style,
        }} 
        className={joinClasses(
          'ava',
          load && 'ava-loader', 
          errorClass, 
          className
        )}
        draggable={!draggable ? false : undefined}
        onError={Err}
        onLoad={Load}
      />
    );
  }
);
