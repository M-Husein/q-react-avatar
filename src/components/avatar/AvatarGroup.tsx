import '../../style/components/AvatarGroup.scss';
import type { ElementType, ReactNode } from 'react';
import { Children } from 'react';
import { joinClasses } from "../../utils";
import { parseSize } from './utils';

export interface AvatarGroupProps {
  As?: ElementType
  className?: string
  maxLength?: number
  size?: number | "xs" | "sm" | "md" | "lg"
  children: ReactNode
}

export function AvatarGroup({
  As = 'span',
  className,
  maxLength = 0,
  size = 'md',
  children,
}: AvatarGroupProps){
  const fixSize = parseSize(size);
  const childs = Children.toArray(children);
  const isMaxLength = !!maxLength && childs.length > maxLength;

  const parseDisplay = (isMaxLength ? childs.slice(0, maxLength) : childs).map((item: any, index: number) => {
    const { children: child, size: s, style, alt, className: cls, ...rest } = item.props;
    const fixAlt = child ? child.props.alt : alt;
    return {
      ...item,
      props: {
        ...rest, // item.props
        size: child ? undefined : size,
        style: {
          ...style,
          marginLeft: index > 0 ? `calc(-${fixSize}px / 2.75)`: null,
        },
        className: child ? joinClasses(child.props.className, cls) : cls,
        alt: fixAlt,
        title: fixAlt,
        children: child ? {
          ...child,
          props: {
            ...child.props,
            size,
          }
        }
        :
        undefined
      }
    }
  });

  return (
    <As
      role="group"
      className={joinClasses("avatar-group", className)}
    >

      {parseDisplay}

      {isMaxLength && (
        <span
          className="ava ava-rest"
          style={{
            width: fixSize,
            height: fixSize,
            fontSize: 'calc(' + fixSize + 'px / 2.25)',
            marginLeft: `calc(-${fixSize}px / 2.75)`
          }}
          title={childs.slice(maxLength).map((item: any) => item.props.children ? item.props.children.props.alt : item.props.alt).join('\n')}
        >
          +{childs.length - maxLength}
        </span>
      )}
    </As>
  );
}
