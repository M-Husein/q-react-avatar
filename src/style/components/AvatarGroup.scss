// 
.avatar-group  {
  display: flex;
  flex-direction: row;
  align-items: stretch;
  user-select: none;
  -webkit-tap-highlight-color: transparent;
}

.avatar-group > * {
  position: relative;
  display: block;
  border: 3.5px solid #fff;
  flex: none;
  text-decoration: none;
}

.avatar-group img.ava {
  display: block;
}

.avatar-group > *:hover {
  z-index: 1;
}

// .avatar-group > .ava + .ava {
//   margin-left: -1.5%;
// }

.ava.ava-rest {
  display: grid;
  place-content: center;
  line-height: 1;
  color: #757a87;
  background-color: #e6e9f3;
  border-radius: 50%;
  font-weight: 600;
}
