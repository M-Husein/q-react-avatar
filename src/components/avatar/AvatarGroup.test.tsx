import { render, screen } from '@testing-library/react';

import { Avatar } from './Avatar';
import { AvatarGroup } from './AvatarGroup';

const IMAGES = [
  { id: 1, name: "Catherine Missal", src: "https://images.unsplash.com/photo-1494790108377-be9c29b29330?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxzZWFyY2h8NHx8dXNlcnxlbnwwfHwwfHw%3D&w=1000&q=80" },
  { id: 2, name: "Siti Umaya", src: "/img/hijab_girl.jpg" },
  { id: 3, name: "Elizabeth Olsen", src: "/img/hijab_girl_2.webp" },
  { id: 4, name: "Jane Doe Taylor", src: "/img/hijab_girl_3.jpg" },
  { id: 5, name: "Bruce Wayne" },
  { id: 6, name: "Angelina Rose", src: "/img/hijab_girl_4.jpg" },
  { id: 7, name: "Diana Prince" },
  { id: 8, name: "Peter Parker" },
];

describe('AvatarGroup component', () => {
  it('AvatarGroup component - should check render with snapshot', () => {
    const component = render(
      <AvatarGroup
        className="avatar-group-snapshot"
      >
        {IMAGES.map((item: any) => <Avatar key={item.id} src={item.src} alt={item.alt} className="rounded-full" />)}
      </AvatarGroup>
    );
    // @ts-ignore
    expect(component).toMatchSnapshot();
  });

  it('renders <AvatarGroup /> default props', () => {
    render(
      <AvatarGroup
        className="avatar-group-default-props"
      >
        {IMAGES.map((item: any) => <Avatar key={item.id} src={item.src} alt={item.alt} className="rounded-full" />)}
      </AvatarGroup>
    );

    // @ts-ignore
    expect(screen.getByRole('group')).toBeInTheDocument();
    // @ts-ignore
    expect(screen.getByRole('group')).toHaveAttribute('class', 'avatar-group avatar-group-default-props');
  });

  it('with maxLength props', () => {
    const { container } = render(
      <AvatarGroup
        maxLength={4}
        className="avatar-group-with-maxLength"
      >
        {IMAGES.map((item: any) => <Avatar key={item.id} src={item.src} alt={item.alt} className="rounded-full" />)}
      </AvatarGroup>
    );

    // @ts-ignore
    expect(screen.getByRole('group')).toBeInTheDocument();
    // @ts-ignore
    expect(container.querySelector('img.ava')).toBeInTheDocument();
    // @ts-ignore
    expect(screen.getByText('+4')).toBeInTheDocument();
  });

  it('with size number props', () => {
    const { container } = render(
      <AvatarGroup
        size={125}
        className="avatar-group-with-size-number"
      >
        {IMAGES.map((item: any) => <Avatar key={item.id} src={item.src} alt={item.alt} className="rounded-full" />)}
      </AvatarGroup>
    );

    // @ts-ignore
    expect(screen.getByRole('group')).toBeInTheDocument();
    // @ts-ignore
    expect(container.querySelector('img.ava[width="125"]')).toBeInTheDocument();
  });

  it('render as div', () => {
    const { container } = render(
      <AvatarGroup
        As="div"
        className="avatar-group-as-div"
      >
        {IMAGES.map((item: any) => <Avatar key={item.id} src={item.src} alt={item.alt} className="rounded-full" />)}
      </AvatarGroup>
    );

    // @ts-ignore
    expect(screen.getByRole('group')).toBeInTheDocument();
    // @ts-ignore
    expect(container.querySelector('img.ava')).toBeInTheDocument();
  });
});
