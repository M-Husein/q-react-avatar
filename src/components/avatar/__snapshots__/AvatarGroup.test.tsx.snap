// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AvatarGroup component > AvatarGroup component - should check render with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="avatar-group avatar-group-snapshot"
        role="group"
      >
        <img
          alt=" "
          class="ava ava-loader rounded-full"
          decoding="async"
          draggable="false"
          height="56"
          loading="lazy"
          src="https://images.unsplash.com/photo-1494790108377-be9c29b29330?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxzZWFyY2h8NHx8dXNlcnxlbnwwfHwwfHw%3D&w=1000&q=80"
          width="56"
        />
        <img
          alt=" "
          class="ava ava-loader rounded-full"
          decoding="async"
          draggable="false"
          height="56"
          loading="lazy"
          src="/img/hijab_girl.jpg"
          width="56"
        />
        <img
          alt=" "
          class="ava ava-loader rounded-full"
          decoding="async"
          draggable="false"
          height="56"
          loading="lazy"
          src="/img/hijab_girl_2.webp"
          width="56"
        />
        <img
          alt=" "
          class="ava ava-loader rounded-full"
          decoding="async"
          draggable="false"
          height="56"
          loading="lazy"
          src="/img/hijab_girl_3.jpg"
          width="56"
        />
        <img
          alt=" "
          aria-label="?"
          class="ava ava-light rounded-full"
          decoding="async"
          draggable="false"
          height="56"
          loading="lazy"
          style="--fs: calc(56px / 2.25); --bg: #5a6268;"
          width="56"
        />
        <img
          alt=" "
          class="ava ava-loader rounded-full"
          decoding="async"
          draggable="false"
          height="56"
          loading="lazy"
          src="/img/hijab_girl_4.jpg"
          width="56"
        />
        <img
          alt=" "
          aria-label="?"
          class="ava ava-light rounded-full"
          decoding="async"
          draggable="false"
          height="56"
          loading="lazy"
          style="--fs: calc(56px / 2.25); --bg: #5a6268;"
          width="56"
        />
        <img
          alt=" "
          aria-label="?"
          class="ava ava-light rounded-full"
          decoding="async"
          draggable="false"
          height="56"
          loading="lazy"
          style="--fs: calc(56px / 2.25); --bg: #5a6268;"
          width="56"
        />
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="avatar-group avatar-group-snapshot"
      role="group"
    >
      <img
        alt=" "
        class="ava ava-loader rounded-full"
        decoding="async"
        draggable="false"
        height="56"
        loading="lazy"
        src="https://images.unsplash.com/photo-1494790108377-be9c29b29330?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxzZWFyY2h8NHx8dXNlcnxlbnwwfHwwfHw%3D&w=1000&q=80"
        width="56"
      />
      <img
        alt=" "
        class="ava ava-loader rounded-full"
        decoding="async"
        draggable="false"
        height="56"
        loading="lazy"
        src="/img/hijab_girl.jpg"
        width="56"
      />
      <img
        alt=" "
        class="ava ava-loader rounded-full"
        decoding="async"
        draggable="false"
        height="56"
        loading="lazy"
        src="/img/hijab_girl_2.webp"
        width="56"
      />
      <img
        alt=" "
        class="ava ava-loader rounded-full"
        decoding="async"
        draggable="false"
        height="56"
        loading="lazy"
        src="/img/hijab_girl_3.jpg"
        width="56"
      />
      <img
        alt=" "
        aria-label="?"
        class="ava ava-light rounded-full"
        decoding="async"
        draggable="false"
        height="56"
        loading="lazy"
        style="--fs: calc(56px / 2.25); --bg: #5a6268;"
        width="56"
      />
      <img
        alt=" "
        class="ava ava-loader rounded-full"
        decoding="async"
        draggable="false"
        height="56"
        loading="lazy"
        src="/img/hijab_girl_4.jpg"
        width="56"
      />
      <img
        alt=" "
        aria-label="?"
        class="ava ava-light rounded-full"
        decoding="async"
        draggable="false"
        height="56"
        loading="lazy"
        style="--fs: calc(56px / 2.25); --bg: #5a6268;"
        width="56"
      />
      <img
        alt=" "
        aria-label="?"
        class="ava ava-light rounded-full"
        decoding="async"
        draggable="false"
        height="56"
        loading="lazy"
        style="--fs: calc(56px / 2.25); --bg: #5a6268;"
        width="56"
      />
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
